lib/devicetrust/enroll/enroll.go

```diff
@@ -55,6 +57,114 @@ func RunCeremony(ctx context.Context, devicesClient devicepb.DeviceTrustServiceC
 	return NewCeremony().Run(ctx, devicesClient, debug, enrollToken)
 }
 
+type RunAdminOutcome int
+
+const (
+	_ RunAdminOutcome = iota // Zero means nothing happened.
+	DeviceEnrolled
+	DeviceRegistered
+	DeviceRegisteredAndEnrolled
+)
+
+// RunAdmin is a more powerful variant of Run: it attempts to register the
+// current device, creates an enrollment token and uses that token to call Run.
+//
+// Must be called by a user capable of performing all actions above, otherwise
+// it fails.
+//
+// Returns the created or enrolled device, an outcome marker and an error. The
+// zero outcome means everything failed.
+//
+// Note that the device may be created and the ceremony can still fail
+// afterwards, causing a return similar to "return dev, DeviceRegistered, err"
+// (where nothing is "nil").
+func (c *Ceremony) RunAdmin(
+	ctx context.Context,
+	devicesClient devicepb.DeviceTrustServiceClient,
+	debug bool,
+) (*devicepb.Device, RunAdminOutcome, error) {
+	// The init message contains the device collected data.
+	init, err := c.EnrollDeviceInit()
+	if err != nil {
+		return nil, 0, trace.Wrap(err)
+	}
+	cdd := init.DeviceData
+	osType := cdd.OsType
+	assetTag := cdd.SerialNumber
+
+	rewordAccessDenied := func(err error, action string) error {
+		if trace.IsAccessDenied(trail.FromGRPC(err)) {
+			log.WithError(err).Debug(
+				"Device Trust: Redacting access denied error with user-friendly message")
+			return trace.AccessDenied(
+				"User does not have permissions to %s. Contact your cluster device administrator.",
```

Comment:
I'm not sure about this message, but I couldn't think of anything better.

If we do nothing, it prints something like this:

```
ERROR: rpc error: code = PermissionDenied desc = access denied to perform action "list" on "device"
```

Reply:
> IMO, for debugging, its best to see actual error message

`tsh -d` prints the original error too, so we're good in that front.

> But for non-debugging (i.e. without explicit --debug), user friendly message are more helpful

👍